Bug fixes for full/wide image alignments.

joemcgill opened the pull request:
## Description
This addresses two image size issues that currently exist with wide/full aligned images using the core/image block:

1. Once an image block is set to wide/full alignments, the inspector controls for adding alternate text or link settings are no longer available.

Commit dbe4042 fixes this issue by including inspector controls even when an image is not resizable (i.e., it's a wide/full aligned image), but conditionally only shows the inspector controls that are related to size settings when an image is, in fact, resizable.

2. If the image size has been manually set to a smaller size (i.e., medium) before changing the alignment to wide/full, the image will continue using the smaller image file as the source url, resulting in blurry, stretched images.

Commit 8651faa addresses this issue by resetting the `url` attribute of the image during `updateAlignment()` to the full size image whenever we're changing the alignment to wide or full.


## How has this been tested?

1. Set an image block to wide or full alignment and see that the alternative text and link settings inspector controls are now displayed.
2. Insert an image, set it to "medium" size, change alignment to "full" and see that the image is not blurry, and is using the full size URL.
3. Run all jest tests after making the changes.

## Screenshots <!-- if applicable -->

## Types of changes

Bug fixes (see above).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

joemcgill committed:
Show inspector controls on images that aren't resizable

This ensures inspector controls for alternate text and link settings are still available whenever an image is not resizable by including inspector controls always, and only conditionally loading specific controls that are relevant to sizing whenever isResizable is true.

joemcgill committed:
Reset image source size when using wide/full alignments.

This resets the `url` attribute of an image to the full size source URL whenever changing alignments to 'wide' or 'full', which prevents a blurry image whenever someone selects a smaller source image before changing to a 'wide' or 'full' alignment.

jasmussen commented:
Nice work, this will be a good fix to have, thank you.

I'm seeing an issue, this is after an npm install && npm run dev:

![issue](https://user-images.githubusercontent.com/1204802/43703521-44499c04-995d-11e8-8739-920986d20011.gif)

This is the full error:

```
react-dom.min.82e21c65.js:144 Uncaught TypeError: Cannot read property 'source_url' of null
    at ImageEdit.updateAlignment (edit.js:175)
    at Object.onClick (index.js:36)
    at Object.onClick (index.js:71)
    at Tooltip.emitToChild (index.js:115)
    at index.js:122
    at Object.Sh (react-dom.min.82e21c65.js:143)
    at Object.invokeGuardedCallback (react-dom.min.82e21c65.js:143)
    at Object.invokeGuardedCallbackAndCatchFirstError (react-dom.min.82e21c65.js:143)
    at Wd (react-dom.min.82e21c65.js:14)
    at Bg (react-dom.min.82e21c65.js:145)
```

This is on the demo content. It could be that this issue is not present when you insert an image fresh.

joemcgill committed:
Sanity check for the source image before resetting the url attribute

This fixes an issue where updating alignments on an image block that doesn't have a related attachment object results in an error.

jasmussen commented:
Nice, this is working for me now. Cool work. Okay from me, though probably need a quick code review from someone.

youknowriad commented on core-blocks/image/edit.js:
```diff
@@ -172,7 +172,7 @@ class ImageEdit extends Component {
 	updateAlignment( nextAlign ) {
 		const extraUpdatedAttributes = [ 'wide', 'full' ].indexOf( nextAlign ) !== -1 ?
-			{ width: undefined, height: undefined } :
+			{ width: undefined, height: undefined, url: this.props.image ? this.props.image.source_url : this.props.attributes.url } :
```
Honestly, I'm not certain this behavior is what we want as some people might still want to use an alternative size (it could be big enough) while using wide/full aligned images. cc @jasmussen

jasmussen replied:
Can you elaborate on what this means?

As far as I'm concerned, if an image is set to _wide_ or _fullwide_, we defer completely to the theme to define what that means.

youknowriad replied:
The `url` used in the `<img>` tag is chosen using an inspector control. If we want to always use the "full sized" image for these, than it's fine, we can keep this change, if we want to allow users to choose which "image size" (including custom registered sizes which can be bign enough), than we shouldn't reset.

jasmussen replied:
Ah, gotcha. So a user uploads a 4000*4000px photo of a badger, then wants to make it _wide_, but they feel 4000*4000 might be too much resolution and chooses a smaller size even if it's wide? Sure, that makes sense to me.

youknowriad replied:
Ok so I took this for a spin and it looks like we're hiding the "Image Size" selector when in full/wide alignments. This means even if you want to change those you can't. which means:

 1- We can ship this PR as: always full image in wide alignedd images
 2- Update the PR to always show the image size selector

I'd appreciate an opinion quickly because I'm trying to get this in for 3.5 :)

youknowriad reviewed (commented):
LGTM overall, just not sure about the image size reset behavior

youknowriad committed:
Allow choosing the image size in wide/full images as well

youknowriad reviewed (approved):
LGTM 👍

mtias commented:
Thanks for these fixes!
